src/pyk/kore/kompiled.py

```diff
@@ -105,6 +105,32 @@ def meet_all_sorts(self, sorts: Iterable[Sort]) -> Sort:
         unit: Sort = SortApp('SortK')
         return reduce(self.meet_sorts, sorts, unit)
 
+    def replace_k_injections_with_ksequence(self, pattern: Pattern) -> Pattern:
+        def _replace_k_injections(pattern: Pattern) -> Pattern:
+            if (
+                type(pattern) is App
+                and pattern.symbol == 'inj'
+                and pattern.sorts[0] != SortApp('SortK')
+                and pattern.sorts[1] == SortApp('SortK')
```

Comment:
This check is confusing to me? Should we instead be saying:

```suggestion
                and pattern.sorts[0] == SortApp('SortKItem')
                and pattern.sorts[1] == SortApp('SortK')
```

Basically very specifically calling out injections from `KItem -> K`, and replacing them?

Or is the thinking here that anything injected into K must go through `KItem` first? 

Maybe we need to ask @dwightguth the design decision behind the sorts `K` and `KItem`. Whenever we have something of sort `B` lower than `K` in place where a `K` is expected, are we expected that we have a `kseq` at the top, then an injection from `B -> KItem` under that?

If so, I guess I would propose doing this pass as:

- A pass that turns injections like `B -> K` into a double injection `B -> KItem -> K` (if `B != KItem` and `B != K`), and then
- A pass that takes injections `KItem -> K` and turns them into `kseq(...)` instead.

Reply:
Okay, I can give you some insight into how KORE is generated by the Java frontend.

The frontend considers KItem a subsort of K, but the backends do not. From the perspective of the backend, `K` as a sort has two constructors and one function. The constructors are `dotk{}`, the empty K sequence, and `kseq{}`, which is the `cons` constructor for the list, and has the following signature:

```
symbol kseq(KItem, K) : K
```

There is also the following symbol:

```
symbol append(K, K) : K
```

This is a fairly standard cons-list-append function.

The frontend converts a list of terms into the backend representation by using `kseq` for each term that is a subsort of KItem and `append` if the term on the left is sort K. This then ultimately gets normalized by the implementation of `append` into a list of `KItem`.

It is never correct to have an injection directly into `K`. The behavior in this case is to create a singleton K sequence whose element is an injection into KItem (unless the term itself is of sort KItem, in which case no injection is needed).

Reply:
The intuition of the K sort is that it's literally just a list of KItems, so it stands to reason that you would need a singleton list if you have a KItem in a place where a K is expected. I don't anticipate that that will change unless the behavior of the K sort is fundamentally changed, which seems unlikely to me since that would break a lot of assumptions semantics are making. If you don't want to support 0 or 2+ element lists, why are you using sort K? And if you do, you need the explicit list representation even if the list is currently of length 1. So I don't expect this behavior will change anytime soon.

Reply:
Hmmmm, OK. In that case, what I would propose is this:

- Run the pass that inserts `kseq` _before_ adding sort injections. Basically look for places where you have a production that expects a `K` under it (like `_==K_`), but the argument is not already of sort `K`, then you add the `KSequence` under that.
- Then run the sort injection pass (which already will correctly infer `KItem` for that position).

This is basically what we were doing before, we were inserting `KSequence` at the KAST level, and letting sort injections happen at Kore level. So I'd say lets go back to that solution. Thanks for the insight @dwightguth !

Reply:
Even if you do this, you still need to use `append` instead of `kseq` if a term in a KSequence other than the last is of sort K when converting from KAST to KORE. Not sure if you're already doing that.

Reply:
You also need to make sure that you flatten any K sequences that are nested /after/ splitting the left and right hand side of a rule if you are creating a new rule or claim.

Reply:
Hmmmm, right now we are not doing any splitting of rules and such, so I don't think that is the concern (this is for individual terms). But we are not emitting `append(...)` in any cases, so that's probably something we need to look at.

I'm wondering how we would even really test that though....

I guess the test which has:

```
KSequence([KVariable('X', 'K'), KVariable('Y', 'K')])
```

Should be emitting something like:

```
append(X:K, Y:K)
```

rather than:

```
kseq(X:K, kseq(Y:K, dotk))
```

And the test that has:

```
KSequence([KVariable('X'), KVariable('Y')])
```

Should be emitting:

```
kseq(X:KItem, Y:K)
```

Does that sound right?

Reply:
Is it important that `kast_to_kore` always produce the same KAST as `kast` (k binary)? If so, could we just build this requirement so it tries both and compares them? I have done some experimenting with this and it seems like there may be other discrepancies. Not sure if they are necessarily important though. For example, `kast_to_kore` outputs
```
App(symbol='Lblbarholder2', sorts=(), args=(App(symbol='inj', sorts=(SortApp(name='SortBaz', sorts=()), SortApp(name='SortBar', sorts=())), args=(EVar(name='VarX', sort=SortApp(name='SortBaz', sorts=())),)), EVar(name='VarX', sort=SortApp(name='SortBaz', sorts=()))))
```
while using `KPrint.kast_to_kore` modified to always use the `kast` binary outputs:
```
App(symbol='Lblbarholder2', sorts=(), args=(App(symbol='inj', sorts=(SortApp(name='SortBaz', sorts=()), SortApp(name='SortBar', sorts=())), args=(EVar(name='VarX', sort=SortApp(name='SortBaz', sorts=())),)), App(symbol='inj', sorts=(SortApp(name='SortBar', sorts=()), SortApp(name='SortBaz', sorts=())), args=(EVar(name='VarX', sort=SortApp(name='SortBar', sorts=())),))))
```
for test `variable-with-multi-sort`.

Reply:
Hmmmm, adding that as a test seems like a good idea! I would first just get this PR finished up (with the ksequence changes, basically going back to how you were doing it at first), then we can make a separate PR which tests all teh `kast_to_kore` conversion data with both techniques and compares. We can just skip tests with `pytest.skip()` for ones that don't work at the moment, and make an issue about it.